### ISSUE TITLE
Add support for Azure resources

### ISSUE THREAD

**Issue by hassankhosseini** [infracost/infracost#64]
Pricing: https://azure.microsoft.com/en-us/pricing/

**Comment by 3mard** [comment]
Not sure if this is doable at the moment

I couldn't find the official APIs for pricing, are we okay we capturing the api used in https://azure.microsoft.com/en-us/pricing/ and then use them as undocumented APIs ?

**Comment by PeterDaveHello** [comment]
There is a Azure Resource RateCard API under preview, might be suitable here:

https://docs.microsoft.com/en-us/azure/cost-management-billing/manage/usage-rate-card-overview

**Comment by 3mard** [comment]
@PeterDaveHello Thanks for the tip:
 But my understanding, the cost management api tells you about what you are already using, meanwhile infracost predicts what your cost will be in the future

**Comment by aliscott** [comment]
> There is a Azure Resource RateCard API under preview

@PeterDaveHello thanks, yeah a RateCard API is what we want but it seems unfortunately this RateCard API has been under preview since 2016 and looks to be somewhat abandoned now.

There seems to be 2 ways I can find for currently getting Azure prices.
1. The JSON files used in the Azure calculators (https://azure.microsoft.com/en-us/pricing/calculator/)
2. The Azure Partner Center rate card api (https://docs.microsoft.com/en-us/partner-center/develop/azure-rate-card-resources)

I think #1 seems the easiest to get started on / prototype with, but at some point we can sign up as a Cloud Solution Provider partners and use #2.

**Comment by PeterDaveHello** [comment]
> @PeterDaveHello Thanks for the tip:
>  But my understanding, the cost management api tells you about what you are already using, meanwhile infracost predicts what your cost will be in the future 
> 

It should also tell you the unit price of the resources, but as @aliscott mentions, there might be some challenges.

**Comment by PeterDaveHello** [comment]
> @PeterDaveHello thanks, yeah a RateCard API is what we want but it seems unfortunately this RateCard API has been under preview since 2016 and looks to be somewhat abandoned now.

Looks like this Azure VM pricing comparison siye is using that API:
https://azureprice.net/About

So it should be working good enough, maybe just not very user-friendlyor something?

**Comment by PeterDaveHello** [comment]
Just notice that there is an Azure Retail Prices API might just be the one needed here!?
https://docs.microsoft.com/en-us/rest/api/cost-management/retail-prices/azure-retail-prices

**Comment by aliscott** [comment]
Thanks everyone for the comments and 👍 s

A quick update: I had a quick look at the Azure PR that’s in progress and it’s looking pretty good. I’ll have time to dive deeper into this at the end of the month so stay tuned.

**Comment by alikhajeh1** [comment]
We completed adding around 300,000 prices from Microsoft Azure to the Cloud Pricing API are are now looking for contributors to help us add [Azure resources](https://github.com/infracost/infracost/issues?q=is%3Aissue+is%3Aopen+label%3Aazure) to Infracost. For anyone interested, please take a look at the [contributing guide](https://github.com/infracost/infracost/blob/master/CONTRIBUTING.md) and join the [community Slack chat](https://www.infracost.io/community-chat) for assistance.

**Comment by hassankhosseini** [comment]
Thank you to everyone who left a +1, a comment and code! We have now added support for more than 65 Azure resources in Infracost, and all is live now in the latest version. We have also added support for Azure DevOps CI/CD!

Keep the votes and issues coming, we'd love to keep iterating and making Infracost better - I'd especially love to see issues around other resources you'd like us to support, or features you'd like to see next.

Cheers!
